promsql/driver_collector.go

```diff
@@ -0,0 +1,626 @@
+
+var (
+	regMu sync.Mutex
+)
+
+func Register(options DriverCollectorOpts) (*DriverCollector, error) {
+	// retrieve the driver implementation we need to wrap with instrumentation
+	db, err := sql.Open(options.DriverName, "")
+	if err != nil {
+		return nil, err
+	}
+	dri := db.Driver()
+	if err = db.Close(); err != nil {
+		return nil, err
+	}
+
+	regMu.Lock()
+	defer regMu.Unlock()
+
+	driverName := options.DriverName + "-collector-"
+	for i := int64(0); i < 1000; i++ {
+		var (
+			found   = false
+			regName = driverName + strconv.FormatInt(i, 10)
+		)
+		for _, name := range sql.Drivers() {
+			if name == regName {
+				found = true
+			}
+		}
+		if !found {
+			driverCollector := Wrap(dri, options)
+			sql.Register(regName, driverCollector)
+			driverCollector.DriverName = regName
+			return driverCollector, nil
+		}
+	}
+	return nil, errors.New("unable to register driver, all slots have been taken")
+}
+
+func Wrap(d driver.Driver, options DriverCollectorOpts) *DriverCollector {
+	return wrapDriver(d, options)
+}
+
+func wrapDriver(d driver.Driver, o DriverCollectorOpts) *DriverCollector {
+	return NewDriverCollector(d, o)
+}
+
+type DriverCollector struct {
+	DriverName string
+	parent     driver.Driver
+	connector  driver.Connector
+
+	// prometheus counters
+	QueryTotalCounter                  prometheus.Counter
+	QuerySuccessfulCounter             prometheus.Counter
+	QueryErroneousCounter              prometheus.Counter
+	TransactionTotalCounter            prometheus.Counter
+	TransactionSuccessfulAmountCounter prometheus.Counter
+	TransactionErroneousAmountCounter  prometheus.Counter
+	ExecutionTotalCounter              prometheus.Counter
+	ExecutionSuccessfulAmountCounter   prometheus.Counter
+	ExecutionErroneousAmountCounter    prometheus.Counter
+}
+
+type DriverCollectorOpts struct {
+	DriverName string
+	Prefix     string
+}
+
+func NewDriverCollector(driver driver.Driver, opts DriverCollectorOpts) *DriverCollector {
+	prefix := opts.Prefix
+	if prefix != "" && !strings.HasSuffix(opts.Prefix, "_") {
+		prefix += "_"
+	}
+
+	return &DriverCollector{
+		parent:                             driver,
+		DriverName:                         opts.DriverName,
+		QueryTotalCounter:                  prometheus.NewCounter(prometheus.CounterOpts{}),
+		QuerySuccessfulCounter:             prometheus.NewCounter(prometheus.CounterOpts{}),
+		QueryErroneousCounter:              prometheus.NewCounter(prometheus.CounterOpts{}),
+		TransactionTotalCounter:            prometheus.NewCounter(prometheus.CounterOpts{}),
+		TransactionSuccessfulAmountCounter: prometheus.NewCounter(prometheus.CounterOpts{}),
+		TransactionErroneousAmountCounter:  prometheus.NewCounter(prometheus.CounterOpts{}),
+		ExecutionTotalCounter:              prometheus.NewCounter(prometheus.CounterOpts{}),
+		ExecutionSuccessfulAmountCounter:   prometheus.NewCounter(prometheus.CounterOpts{}),
+		ExecutionErroneousAmountCounter:    prometheus.NewCounter(prometheus.CounterOpts{}),
```

Comment:
I think you should be adding names and descriptions for those metrics.

Remember adding the prefix.